drivers/spi/spi_stm32.c

```diff
@@ -0,0 +1,584 @@
+/*
+ * Copyright (c) 2016 BayLibre, SAS
+ * Copyright (c) 2018 Bobby Noelte
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** @file
+ * @brief STM32 SPI implementation
+ * @defgroup device_driver_spi_stm32 STM32 SPI inplementation
+ *
+ * A common driver for STM32 SPI. SoC specific adaptions are
+ * done by device tree and soc.h.
+ *
+ * @{
+ */
+
+/** Compile if at least one device is activated
+ * @code{.codegen}
+ * compatible = 'st,stm32-spi-fifo'
+ * codegen.if_device_tree_controller_compatible('SPI', compatible)
```

Comment:
This doesn't look right.

I thought that drivers had to declare themselves and their IRQs (by calling DEVICE_AND_API_INIT and friends).

But as this is written, the generation of the DEVICE_AND_API_INIT calls will be omitted when there is no activated and compatible SPI controller.

Will this driver be valid/useful when this codegen guard is false?

If not, then we shouldn't be compiling this file at all. But I can't see any mechanism that is preventing this,
if there was, this codegen guard wouldn't be necessary.

Reply:
codegen.if_device_tree_controller_compatible('SPI', compatible) stops code generation if there is no SPI controller that is compatible to 'st,stm32-spi-fifo'. 

Looking for a better function name - but it is documented.